### PR TITLE
fix(bpf): Handle the post-6.4 definition of `iov_iter`

### PR DESCRIPTION
**Pull request by vadorovsky**
Kernel v6.4 introduced a change in the layout of `struct iov_iter` (torvalds/linux@747b1f6). Handle it accordingly in `on_socket_sendmsg` and `on_socket_recvmsg` BPF programs.

Fixes: #194

### CODE REVIEW

**Comment by vadorovsky** [comment]
Marking as a draft. It fixes the issue on >=6.4 kernels, but I need to come up with a proper way to handle older ones. I will probably have to handle that with checking `bpf_core_field_exists` and then doing manual offsets and pointer arithmetics for older kernels. It's gonna be fun. :smiling_face_with_tear: 

@banditopazzo can you confirm that it works for your new kernel as well?

**Comment by vadorovsky** [comment]
Interesting idea with the "compat" header - I think we could try putting there only the old `struct iov_iter` definition (under a different name, then try to treat the pointer as either `struct iov_iter` or `struct iov_iter_compat` with `bpf_core_field_exists` checks, determining which definition suits. Ugly, but should be somehow safe (safer than anything else :v).

**Comment by vadorovsky** [comment]
@banditopazzo @krsh Ready for review. I went with the approach of having a compat header, described above. And I also made an xtask for building the kernel and generating vmlinux.h.

The OpenSSL error is gone after switching to rustls.

There is still some crate failing to build on riscv, I'm trying to fix that.

**Comment by vadorovsky** [comment]
@banditopazzo Can you take a look?

File: crates/modules/network-monitor/probes.bpf.c [hunk 1]
```diff
@@ -312,6 +313,27 @@ static __always_inline void on_accept_exit(void *ctx, long ret) {
...
+  // Definition of `struct iov_iter` used in new kernels (>=6.4).
+  const struct iov_iter *msg_iter_nocompat = msg_iter;
+
+  // Check if `msg_iter` matches the new definition of `struct iov_iter`. The
+  // `__ubuf_iovec` field doesn't exist in kernels <= 6.4.
+  if (bpf_core_field_exists(msg_iter_nocompat->__ubuf_iovec)) {
+    if (BPF_CORE_READ(msg_iter_nocompat, iter_type) == ITER_UBUF) {
+      return &msg_iter_nocompat->__ubuf_iovec;
+    }
+    return BPF_CORE_READ(msg_iter_nocompat, __iov);
```
Comment (banditopazzo) [inline crates/modules/network-monitor/probes.bpf.c:329]: why in the first case the access is plain and only in the other case you use `BPF_CORE_READ`?

Comment (vadorovsky) [reply crates/modules/network-monitor/probes.bpf.c:329]: `__ubuf_iovec` and `__iov` have different types, the former being `struct iovec` and the latter - `struct *iovec`. Doing `&BPF_CORE_READ` is impossible.

But thanks for pointing it out. I probably need to think of some better way of handling these two cases. And return a pointer only when we are supposed to use `__iov`, not `__ubuf_iovec`. I'll think about it.

Comment (vadorovsky) [reply crates/modules/network-monitor/probes.bpf.c]: Actually, I found a clean solution - returning `iov_iter` directly in that new function solves the problem entirely (I can just do `BPF_CORE_READ(msg_iter_nocompat, __ubuf_iovec.iov_base)`).

This is how I'm doing it now (I pushed the change):

https://github.com/Exein-io/pulsar/blob/1f53a1916aa93d0d1c312e35f6f2de5d47cc7209/crates/modules/network-monitor/probes.bpf.c#L318-L335

Let me know if you agree with that.

Comment (banditopazzo) [reply crates/modules/network-monitor/probes.bpf.c]: yeah, now it looks great.

When I saw the arrow operator on the kernel type, it raised me the alarm. I didn't even notice the different types

**Review by banditopazzo — APPROVED** [review]

**Pull request merged**